### ISSUE TITLE
About visualizing the sfm trajectory and localization results

### ISSUE THREAD

**Issue by oym1994** [cvg/Hierarchical-Localization#102]
Hi, thanks for your great work.

As far I have successfully reconstructed the model and localize the query image.Unfortunately some strange things happened:
1, When I use the colmap gui to see the model, it seems right. But when using colmap gui to export the model to .txt file,  reading the pose from images.txt and then publish the pose for rviz visualization, the trajectory is very strange.

2,On the other hand, each query image has more than 30 inlier matches after PnP localization, while visualizing in rviz, the trajectory is also very strange. I clearly know the related output file format for the results.

3, before reconstruction, a calibrated camera intrinsic(OPENCV model) is given, but the intrinsic will be largely modified by the sfm process, what would be the problem?
  **calibrated camera intrinsic**  [581.044540 581.000000 296.00000 248.000000 0.015000 -0.004700 0.001380 0.000186]
  **Modified camera intrinsic**    [531.385 371.463 296 248 -0.0153394 -0.016669 0.00981128 0.00519208]

**Reconstructed model shown in colmap gui**
![Screenshot from 2021-09-27 15-02-09](https://user-images.githubusercontent.com/23044337/134880393-cc9253bf-7850-423e-ace7-552e3477c7aa.png)

**sfm trajectory in rviz**

**query image pose in rviz**

**Comment by lasuomela** [comment]
Did you check the [Colmap coordinate system reference](https://colmap.github.io/format.html#text-format)? The translation- and rotation vectors shown in Colmap don't correspond to a pose with location T and orientation R in a world coordinate system. Instead,

"The reconstructed pose of an image is specified as the projection from world to the camera coordinate system of an image using a quaternion (QW, QX, QY, QZ) and a translation vector (TX, TY, TZ). The quaternion is defined using the Hamilton convention, which is, for example, also used by the Eigen library. The coordinates of the projection/camera center are given by -R^t * T, where R^t is the inverse/transpose of the 3x3 rotation matrix composed from the quaternion and T is the translation vector. The local camera coordinate system of an image is defined in a way that the X axis points to the right, the Y axis to the bottom, and the Z axis to the front as seen from the image."

TLDR; you might have to transform the projection parameters of Colmap into the [ROS rep-103 format](https://www.ros.org/reps/rep-0103.html#coordinate-frame-conventions).

**Comment by oym1994** [comment]
Hi, thanks for your immediate response. You are right. After following your advice, the sfm trajectory seems to be right. But the trajectory of query images is still strange. It seems that the algorithm cant produce enough precise localisation result as far on my own datasets with the default parameters(superpoint 1000 + netVLAD + superglue). How should I do to improve the localisation precision? 

Another thing is how to explain the 3rd question: before reconstruction, a calibrated camera intrinsic(OPENCV model) is given, but the intrinsic will be largely modified by the sfm process, from [581.044540 581.000000 296.00000 248.000000 0.015000 -0.004700 0.001380 0.000186] to [531.385 371.463 296 248 -0.0153394 -0.016669 0.00981128 0.00519208], especially fy.  Can this change mean that the sfm model would be not precise?

Below shows the sfm trajectory(red) and query images trajectory(blue). Ideally these two trajectories should be close and smooth enough. 
![compare](https://user-images.githubusercontent.com/23044337/135019506-59042cb1-7d36-4b04-816b-03fd5468515d.png)

**Comment by Skydes** [comment]
This large change means either that 1) the initial calibration is inaccurate, or 2) the configuration of the scene and the cameras is such that fy is not well constrained during the optimization. If you believe that the calibration is accurate, then you might want to freeze the intrinsics during the reconstruction.
